raster-foundry/raster-foundry
# Improve footprint and task grid calculations

## Overview

This PR:

- calculates a more complex footprint for scenes that go through geotiff processing
- calculates a more complex task grid if the user doesn't post a geometry to the grid endpoint

## Notes

~I didn't do the annotate testing because my local annotate hits the login loop on both chrome and firefox right now~  :sob:don't need annotate for testing it turns out :sunglasses: 

EASIER CODE REVIEW:

https://github.com/raster-foundry/raster-foundry/pull/5266/files/b8209f4e6c89e923855b5639ab38522c75063a55

I noticed some inconsistency with whether I'd remembered to format after saving in my changes, so I `black`'ed everything, but that makes the code changes impossible to read.

### Checklist

- [x] Any new SQL strings have tests (exercised by existing grid test + generator update -- expecting them to fail)

## Testing Instructions

- assemble api
- bring up your server + frontend
- get The Bad Tif :tm: (I'll send it to you via s3)
- upload it into a project
- process that upload
- mouse over the scene in the project scenes list to observe your super cool footprint with holes and weird edges
- get a token from the frontend and export is as `RF_TOKEN`
- send a task grid request for your project with a nulled out geometry, like this:
  - put this json in `post.json`
```json
{
  "type": "Feature",
  "properties": {
    "xSizeMeters": "25",
    "ySizeMeters": "25"
  },
  "geometry": null,
  "projectId": "YOUR-PROJECT-ID",
  "layerId": "YOUR-PROJECT'S-DEFAULT-LAYER-ID"
}
```
  - `cat post.json | http :9091/api/projects/YOUR-PROJECT-ID/layers/YOUR-PROJECT'S-DEFAULT-LAYER-ID/tasks/grid Authorization:$RF_TOKEN`
- get a feature collection of your task with: `http :9091/api/projects/YOUR-PROJECT-ID/layers/YOUR-PROJECT'S-DEFAULT-LAYER-ID/tasks Authorization:$RF_TOKEN> features.json`
- copy `features.json` into `geojson.io` and see that the task grid follows the weird footprint

- it should succeed
- request the task grid 

Closes raster-foundry/annotate#461
